File: drs-go/processor.go
package drs

import (
	"fmt"
	"log"
	"runtime/debug"

	"github.com/ironbay/dynamic"
	"github.com/ironbay/go-util/console"
	"github.com/streamrail/concurrent-map"
)

type Message struct {
	Conn    *Connection
	Command *Command
	Context map[string]interface{}
}

type Processor struct {
	handlers map[string][]func(*Message) (interface{}, error)
	pending  cmap.ConcurrentMap
	stats    cmap.ConcurrentMap

	Before func(msg *Message)
	After  func(msg *Message, result interface{}, err error)
}

type Stats struct {
	Errors     int64 `json:"errors"`
	Exceptions int64 `json:"exceptions"`
	Success    int64 `json:"success"`
}

func NewProcessor() *Processor {
	return &Processor{
		handlers: make(map[string][]func(*Message) (interface{}, error)),
		pending:  cmap.New(),
		stats:    cmap.New(),
		Before:   func(msg *Message) {},
		After:    func(msg *Message, result interface{}, err error) {},
	}
}

func (this *Processor) Enqueue(key string) chan *Command {
	block := make(chan *Command)
	this.pending.Set(key, block)
	return block
}

func (this *Processor) On(action string, cb ...func(*Message) (interface{}, error)) {
	this.handlers[action] = cb
}

func (this *Processor) Off(action string) {
	delete(this.handlers, action)
}

func (this *Processor) Process(conn *Connection, cmd *Command) {
	if cmd.Action == ERROR || cmd.Action == EXCEPTION || cmd.Action == RESPONSE {
		match, ok := this.pending.Get(cmd.Key)
		if !ok {
			return
		}
		this.pending.Remove(cmd.Key)
		match.(chan *Command) <- cmd
		return
	}
	// if this.parent != nil {
	// 	this.parent.Process(conn, cmd)
	// 	return
	// }
	resp, err := this.Invoke(conn, cmd)
	conn.respond(cmd.Key, resp, err)
}

func (this *Processor) Invoke(conn *Connection, cmd *Command) (result interface{}, err error) {
	message := &Message{
		Conn:    conn,
		Command: cmd,
		Context: dynamic.Empty(),
	}
	defer func() {
		this.After(message, result, err)
	}()
	defer func() {
		if r := recover(); r != nil {
			log.Println(r)
			console.JSON(cmd)
			log.Println(string(debug.Stack()))
			var ok bool
			if err, ok = r.(error); !ok {
				err = Exception(fmt.Sprint(r))
			}
		}
	}()
	this.Before(message)
	handlers := this.handlers[cmd.Action]
	if handlers == nil {
		return nil, Error("No handlers for " + cmd.Action)
	}
	for _, cb := range handlers {
		result, err = cb(message)
		if err != nil {
			return nil, err
		}
	}
	return result, nil
}

func (this *Processor) Clear() {
	for kv := range this.pending.Iter() {
		kv.Val.(chan *Command) <- &Command{
			Key:    kv.Key,
			Action: ERROR,
			Body: dynamic.Build(
				"message", "Connection closed",
			),
		}
		this.pending.Remove(kv.Key)
	}
}

File: drs-ts/src/processor.ts
import Connection from './connection'
import Command from './command'

declare type CommandHandler = (Connection, Command, Object) => Object

class Processor {
	private _handlers: Map<string, Array<CommandHandler>>
	private _pending: Map<string, (cmd: Command) => void>
	public redirect: Processor

	constructor() {
		this._pending = new Map<string, (cmd: Command) => void>()
		this._handlers = new Map<string, Array<CommandHandler>>()
	}

	public async on(action: string, ...handlers: Array<CommandHandler>) {
		this._handlers[action] = handlers
	}

	public async wait(cmd: Command, cb: () => void): Promise<Object> {
		const result  = <Command>(await new Promise(async (resolve) => {
			await cb()
			this._pending[cmd.key] = resolve
		}))
		if (result.action !== 'drs.response')
			throw result.body
		return result.body
	}

	protected async process(cmd: Command, conn: Connection) {
		if (this.redirect)
			return this.redirect.process(cmd, conn)
		if (cmd.action === 'drs.response' || cmd.action === 'drs.error' || cmd.action === 'drs.exception') {
			const match = this._pending[cmd.key]
			if (match)
				match(cmd)
			return
		}

		const handlers = this._handlers[cmd.action]
		if (!handlers)
			return
		let context = {}
		const output: Command = {
			key: cmd.key,
			action: 'drs.response',
			body: {},
		}
		try {
			for (let h of handlers) {
				cmd.body = await h(cmd, conn, context)
			}
		} catch (ex) {
			cmd.action = 'drs.error'
			cmd.body = ex
		}
		conn.fire(cmd)
	}

	protected async clear() {
		for (let key in this._pending) {
			this._pending[key]({
				action: 'drs.exception',
				body: {
					message: 'Connection closed'
				}
			})
		}
	}
}

export default Processor
